include/simdjson/padded_string.h

```diff
@@ -33,77 +46,156 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
-      data_ptr[length] = '\0'; // easier when you need a c_str
+  {
+    reset(length);
+    if (data_ptr != nullptr) {
+      // done in reset() -> data_ptr[length] = '\0';
+    }
+    else {
+      /* do what? */
+    }
   }
 
-  explicit padded_string(char *data, size_t length) noexcept
-      : viable_size(length), data_ptr(allocate_padded_buffer(length)) {
+  explicit padded_string(const char *data, size_t length) noexcept
+      : padded_string(length) 
+  {
     if ((data != nullptr) and (data_ptr != nullptr)) {
       memcpy(data_ptr, data, length);
-      data_ptr[length] = '\0'; // easier when you need a c_str
+      // probably not necessary
+      data_ptr[length] = '\0'; 
+    } else {
+      /* do what? */
     }
   }
 
-  // note: do not pass std::string arguments by value
-  padded_string(const std::string & str_ ) noexcept
-      : viable_size(str_.size()), data_ptr(allocate_padded_buffer(str_.size())) {
-    if (data_ptr != nullptr) {
-      memcpy(data_ptr, str_.data(), str_.size());
-      data_ptr[str_.size()] = '\0'; // easier when you need a c_str
-    }
+    // note: do pass std::string_view arguments by value
+  padded_string(std::string_view sv_) noexcept
+      : padded_string(sv_.data(), sv_.size()) {}
+
+  // move ctor
+  padded_string(padded_string &&other_) noexcept {
+    // calling move assignment
+    *this = std::move(other_);
   }
 
-  // note: do pass std::string_view arguments by value
-  padded_string(std::string_view sv_) noexcept
-      : viable_size(sv_.size()), data_ptr(allocate_padded_buffer(sv_.size())) {
-    if (data_ptr != nullptr) {
-      memcpy(data_ptr, sv_.data(), sv_.size());
-      data_ptr[sv_.size()] = '\0'; // easier when you need a c_str
-    }
+  // move assignment
+  padded_string &operator=(padded_string && other_ ) {
+    using namespace std;
+    std::swap(data_ptr, other_.data_ptr);
+    std::swap(viable_size, other_.viable_size);
+    // free the payload
+    simdjson::aligned_free_char(other_.data_ptr);
+    other_.data_ptr = nullptr;
+    other_.viable_size = 0;
+    return *this;
   }
 
-  padded_string(padded_string &&o) noexcept
-      : viable_size(o.viable_size), data_ptr(o.data_ptr) {
-    o.data_ptr = nullptr; // we take ownership
+  // load from file to padded_string
+  // reset the ps_
+  // return the result of fread()
+  // return 0 on error
+  static errno_t load(padded_string &ps_, FILE *fp) noexcept {
+
+    if (std::fseek(fp, 0, SEEK_END) < 0) {
+      std::fclose(fp);
+      return errno;
+    }
+    long llen = std::ftell(fp);
+    if ((llen < 0) || (llen == LONG_MAX)) {
+      std::fclose(fp);
+      return errno;
+    }
+
+    // now make the temporary p string
+    size_t length_ = (size_t)llen;
+
+    if (!ps_.reset(length_))
+      return ENOMEM;
+
+    std::rewind(fp);
+    size_t readb = std::fread(ps_.data(), 1, length_, fp);
+    std::fclose(fp);
+
+    if (readb != length_) {
+      // free mem before bad return
+      ps_.~padded_string();
```

Comment:
I am really uneasy about explicit calls to the destructor. I don't think it is idiomatic.

Reply:
removed already